nbqa/handle_magics.py

```diff
@@ -210,4 +156,192 @@ class CellMagicHandler(MagicHandler):
+
+    @staticmethod
+    def _get_syntax_tree(stmt: str) -> Optional[AST]:
+        """Return the input statement parsed as AST.
+
+        Parameters
+        ----------
+        stmt : str
+            String to be parsed by `ast.parse`
+
+        Returns
+        -------
+        Optional[AST]
+            Return AST if the statement is a valid python code else None
+        """
+        syntax_tree: Optional[AST] = None
+        with contextlib.suppress(SyntaxError):
+            syntax_tree = ast.parse(stmt)
+
+        return syntax_tree
+
+    @staticmethod
+    def _contains_callable(syntax_tree: AST) -> bool:
+        """Return True if the syntax tree contains node of type `ast.Call`.
+
+        Parameters
+        ----------
+        syntax_tree
+            Python code parsed as Abstract Syntax Tree
+
+        Returns
+        -------
+        bool
+            True if the python code contains a callable
+        """
+        return any(isinstance(node, ast.Call) for node in ast.walk(syntax_tree))
+
+    def replace_magic(self) -> str:
+        """
+        Return python code to be replace the input ipython magic.
+
+        Returns
+        -------
+        str
+            Python code to replace the ipython magic
+        """
+        # strip %line_magic from the line magic statement
+        code: str = re.sub(r"%\w+", "", self._ipython_magic).strip()
+        syntax_tree = self._get_syntax_tree(code)
+        if syntax_tree and self._contains_callable(syntax_tree):
+            self._contains_code = True
+            return self._MAGIC_TEMPLATE_WITH_CODE.format(code=code, token=self._token)
+
+        return super().replace_magic()
+
+    def restore_magic(self, cell_source: str) -> str:
+        """Return the cell source with the ipython magic restored.
+
+        Parameters
+        ----------
+        cell_source : str
+            Source code of the notebook cell
+
+        Returns
+        -------
+        str
+            Cell source with ipython magic restored.
+        """
+        if self._contains_code:
+            pattern = MagicHandler._get_regex_pattern(
+                self._MAGIC_WITH_CODE_REGEX_TEMPLATE, self._token
+            )
+            return self._restore_magic_with_modified_code(pattern, cell_source)
+
+        return super().restore_magic(cell_source)
+
+    def _restore_magic_with_modified_code(
+        self, replacement_code_pattern: Pattern[str], cell_source: str
+    ) -> str:
+        """Return cell source code with ipython magic statement restored.
+
+        Parameters
+        ----------
+        replacement_code_pattern : Pattern[str]
+            Regex pattern to find the python code replacing the ipython magic
+        cell_source : str
+            Source code of the notebook cell
+
+        Returns
+        -------
+        str
+            Cell source with ipython magic restored
+        """
+        ipython_magic = self._ipython_magic
+        with contextlib.suppress(AssertionError):
```

Comment:
This method needs refactoring. Any better way to write without using if not None checks?

Reply:
@MarcoGorelli: Do you see any drawback in using assert for testing `if something is not None` ?

Reply:
Is the assert just there to check that there was a match? If so, Anthony Sottile has a video on some options: https://youtu.be/yH6L3UiK9WM

Reply:
> Is the assert just there to check that there was a match? 

Yes. Thanks I will check the video.

Reply:
Thanks @MarcoGorelli. The youtube video you shared was really helpful.

Reply:
no worries - we don't need to suppress the assertionerror any more though, do we?

Reply:
I wanted to catch the `AssertionError` and return a default value instead of throwing it to the user. What are  your thoughts?

Reply:
Part of me thinks that yes, we should throw the error back to the user, along with an encouragement to report a bug - what do you think?

Reply:
IMHO it is nicer if a tool annoys me with spammy warnings than just exploding 😝

Reply:
> Part of me thinks that yes, we should throw the error back to the user, along with an encouragement to report a bug - what do you think?

I am scared of this approach. Imagine if many such bugs are reported at the same time. I agree with your idea of encouraging the users to report such bugs. Nbqa has some projects already in pre-commit workflows. If I raise this error, it won't look good from our part right?

@s-weigand: I like your idea, but first is this what you are saying? Raise some warning from nbqa (write some message to stderr ) when we hit AssertionError?

So taking the middle ground -> I can raise some warnings from nbqa when AssertionError is raised and display well crafted warnings to the user with message to file bug report. But nbqa will handle this error gracefully and will proceed with further notebook processing.

Reply:
OK, thanks both for your input, let's go with a warning then :+1: and leave the return code intact

Reply:
I updated the PR to raise a warning.